liwords-ui/src/shared/player_avatar.tsx

```diff
@@ -1,16 +1,114 @@
-import React from 'react';
+import React, { useEffect } from 'react';
 import { fixedCharAt } from '../utils/cwgame/common';
 import './avatar.scss';
-import { Tooltip } from 'antd';
+import axios from 'axios';
+import { toAPIUrl } from '../api/api';
+import { useMountedState } from '../utils/mounted';
+import { notification, Tooltip, Modal, Alert, Button, Upload } from 'antd';
+import { EditOutlined, UploadOutlined } from '@ant-design/icons';
 import { PlayerMetadata } from '../gameroom/game_info';
 const colors = require('../base.scss');
 
 type AvatarProps = {
   player: Partial<PlayerMetadata> | undefined;
   withTooltip?: boolean;
+  editable?: boolean;
 };
 
 export const PlayerAvatar = (props: AvatarProps) => {
+  const { useState } = useMountedState();
+
+  const [updateModalVisible, setUpdateModalVisible] = useState(false);
+  const [avatarErr, setAvatarErr] = useState('');
+  const [avatarUrl, setAvatarUrl] = useState<string | undefined>('');
+  const [avatarFile, setAvatarFile] = useState(new File([''], ''));
+
+  useEffect(() => {
+    setAvatarUrl(props.player?.avatar_url);
+  }, [props.player]);
+
+  useEffect(() => {
+    setAvatarErr('');
+    var fileInput = document.getElementById(
+      'avatar-file-input'
+    ) as HTMLInputElement;
+    if (fileInput !== null) {
+      fileInput.value = '';
+    }
+  }, [updateModalVisible]);
+
+  var okButtonDisabled = avatarFile == null || avatarFile.name.length === 0;
+  const fileProps = {
+    beforeUpload: (file: File) => {
+      return false;
+    },
+    maxCount: 1,
+    onChange: (info: any) => {
+      if (info.fileList.length > 0) {
+        setAvatarFile(info.fileList.slice(-1)[0].originFileObj);
+      } else {
+        setAvatarFile(new File([''], ''));
+      }
+    },
+    accept: 'image/jpeg',
+    showUploadList: false,
+  };
+
+  const updateModal = (
+    <Modal
+      className="avatar-update-modal"
+      title="Update avatar"
+      visible={updateModalVisible}
+      okText="Upload"
+      okButtonProps={{ disabled: okButtonDisabled }}
+      onCancel={() => {
+        setUpdateModalVisible(false);
+      }}
+      onOk={() => {
+        console.log(avatarFile);
+        var reader = new FileReader();
+        reader.onload = function () {
```

Comment:
using an arrow function in general is a bit better `reader.onload = () => { ...}`